core/drivers/crypto/caam/caam_jr.c

```diff
@@ -0,0 +1,676 @@
+
+	/* Ensure that physical memory is up to date */
+	cache_operation(TEE_CACHECLEAN, cur_inrings,
+			sizeof(struct inring_entry));
+
+	/*
+	 * Increment index to next JR input entry taking care that
+	 * it is a circular buffer of nb_jobs size.
+	 */
+	jr_privdata->inwrite_index++;
+	jr_privdata->inwrite_index %= jr_privdata->nb_jobs;
+
+	/* Ensure that input descriptor is pushed in physical memory */
+	cache_operation(TEE_CACHECLEAN, jobctx->desc,
+			DESC_SZBYTES(caam_desc_get_len(jobctx->desc)));
+
+	/* Inform HW that a new JR is available */
+	caam_hal_jr_add_newjob(jr_privdata->baseaddr);
+
+	*job_id = job_mask;
+	retstatus = CAAM_NO_ERROR;
+
+end_enqueue:
+	cpu_spin_unlock_xrestore(&jr_privdata->inlock, exceptions);
+
+	return retstatus;
+}
+
+/*
+ * Synchronous job completion callback
+ *
+ * @jobctx   Job context
+ */
+static void job_done(struct caam_jobctx *jobctx)
+{
+	jobctx->completion = true;
+}
+
+void caam_jr_cancel(uint32_t job_id)
+{
+	unsigned int idx = 0;
+
+	JR_TRACE("Job cancel 0x%" PRIx32, job_id);
+	for (idx = 0; idx < jr_privdata->nb_jobs; idx++) {
+		/*
+		 * Search for the caller information corresponding to
+		 * the job_id mask.
+		 */
+		if (jr_privdata->callers[idx].job_id == job_id) {
+			/* Clear the Entry Descriptor */
+			jr_privdata->callers[idx].pdesc = 0;
+			jr_privdata->callers[idx].job_id = JR_JOB_FREE;
+			return;
+		}
+	}
+}
+
+enum caam_status caam_jr_dequeue(uint32_t job_ids, unsigned int timeout_ms)
+{
+	uint32_t job_complete = 0;
+	uint32_t nb_loop = 0;
+	bool infinite = false;
+
+	if (timeout_ms == UINT_MAX)
+		infinite = true;
+	else
+		nb_loop = timeout_ms * 100;
+
+	do {
+		/* Call the do_jr_dequeue function to dequeue the jobs */
+		job_complete = do_jr_dequeue(job_ids);
+
+		if (job_complete & job_ids)
+			return CAAM_NO_ERROR;
+
+		/* Check if JR interrupt otherwise wait a bit */
+		if (!caam_hal_jr_check_ack_itr(jr_privdata->baseaddr))
+			caam_udelay(10);
+	} while (infinite || (nb_loop--));
+
+	return CAAM_TIMEOUT;
+}
+
+enum caam_status caam_jr_enqueue(struct caam_jobctx *jobctx, uint32_t *job_id)
+{
+	enum caam_status retstatus = CAAM_FAILURE;
+	__maybe_unused int timeout  = 10; /* Nb loops to pool job completion */
+
+	if (!jobctx)
+		return CAAM_BAD_PARAM;
+
+	JR_DUMPDESC(jobctx->desc);
+
+	if (!jobctx->callback && job_id) {
+		JR_TRACE("Job Callback not defined whereas asynchronous");
+		return CAAM_BAD_PARAM;
+	}
+
+	if (jobctx->callback && !job_id) {
+		JR_TRACE("Job Id not defined whereas asynchronous");
+		return CAAM_BAD_PARAM;
+	}
+
+	jobctx->completion = false;
+	jobctx->status = 0;
+
+	/*
+	 * If parameter job_id is NULL, the job is synchronous, hence use
+	 * the local job_done callback function
+	 */
+	if (!jobctx->callback && !job_id) {
+		jobctx->callback = job_done;
+		jobctx->context = jobctx;
+	}
+
+	retstatus = do_jr_enqueue(jobctx, &jobctx->id);
+
+	if (retstatus != CAAM_NO_ERROR) {
+		JR_TRACE("enqueue job error 0x%08" PRIx32, retstatus);
```

Comment:
s/`%08" PRIx32`/`%08x"`/  
(`retstatus` is an enum and will evaluate as an `int` or `unsigned int`, not explicitly a 32bit value)

Reply:
ok